microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/FailsafeChainedFuture.java

```diff
@@ -54,21 +55,36 @@ public boolean isDone() {
 
     @Override
     @SuppressWarnings("unchecked")
-    public T get() throws InterruptedException, ExecutionException {
-        final T result = delegate.get();
-        if (result instanceof Future) {
-            return ((Future<T>) result).get();
+    public T get() throws CancellationException, InterruptedException, ExecutionException {
+        try {
+            final T result = delegate.get();
+            if (result instanceof Future) {
+                return ((Future<T>) result).get();
+            }
+            return result;
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof CancellationException) {
+                throw (CancellationException) e.getCause();
+            }
+            throw e;
         }
-        return result;
     }
 
     @Override
     @SuppressWarnings("unchecked")
-    public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        final T result = delegate.get();
-        if (result instanceof Future) {
-            return ((Future<T>) result).get(timeout, unit);
+    public T get(long timeout, TimeUnit unit) throws CancellationException, InterruptedException,
```

Comment:
This is begging for refactoring because the two methods are nearly identical. I wondered about creating a third method `get(Supplier<T>)` that would accept a different lambda from each of the existing `get` methods. Then I remembered (admittedly after playing around a bit in an IDE) that the exceptions thrown from the new `get` would mandate a try/catch block in the two lambda expressions so we'd end up not saving much. I remember bumping into this a few times and coming away frustrated each time.

Reply:
I'll try to refactor it.